runtime/src/registrar.rs

```diff
@@ -0,0 +1,1346 @@
+		/// `code` and `initial_head_data` are used to initialize the parathread's state.
+		fn register_parathread(origin,
+			code: Vec<u8>,
+			initial_head_data: Vec<u8>
+		) {
+			let who = ensure_signed(origin)?;
+
+			T::Currency::reserve(&who, T::ParathreadDeposit::get())?;
+
+			let info = ParaInfo {
+				scheduling: Scheduling::Dynamic,
+			};
+			let id = <Self as Registrar<T::AccountId>>::new_id();
+
+			let _ = <Self as Registrar<T::AccountId>>::
+				register_para(id, info, code, initial_head_data);
+
+			<Debtors<T>>::insert(id, who);
+
+			Self::deposit_event(Event::ParathreadRegistered(id));
+		}
+
+		/// Place a bid for a parathread to be progressed in the next block.
+		///
+		/// This is a kind of special transaction that should by heavily prioritized in the
+		/// transaction pool according to the `value`; only `ThreadCount` of them may be presented
+		/// in any single block.
+		fn select_parathread(origin,
+			#[compact] _id: ParaId,
+			_collator: CollatorId,
+			_head_hash: T::Hash
+		) {
+			ensure_signed(origin)?;
+			// Everything else is checked for in the transaction `SignedExtension`.
+		}
+
+		/// Deregister a parathread and retrieve the deposit.
+		///
+		/// Must be sent from a `Parachain` origin which is currently a parathread.
+		///
+		/// Ensure that before calling this that any funds you want emptied from the parathread's
+		/// account is moved out; after this it will be impossible to retrieve them (without
+		/// governance intervention).
+		fn deregister_parathread(origin) {
+			let id = parachains::ensure_parachain(<T as Trait>::Origin::from(origin))?;
+
+			let info = Paras::get(id).ok_or("invalid id")?;
+			if let Scheduling::Dynamic = info.scheduling {} else { Err("invalid parathread id")? }
+
+			<Self as Registrar<T::AccountId>>::deregister_para(id)?;
+			Self::force_unschedule(|i| i == id);
+
+			let debtor = <Debtors<T>>::take(id);
+			let _ = T::Currency::unreserve(&debtor, T::ParathreadDeposit::get());
+
+			Self::deposit_event(Event::ParathreadRegistered(id));
+		}
+
+		/// Swap a parachain with another parachain or parathread. The origin must be a `Parachain`.
+		/// The swap will happen only if there is already an opposite swap pending. If there is not,
+		/// the swap will be stored in the pending swaps map, ready for a later confirmatory swap.
+		///
+		/// The `ParaId`s remain mapped to the same head data and code so external code can rely on
+		/// `ParaId` to be a long-term identifier of a notional "parachain". However, their
+		/// scheduling info (i.e. whether they're a parathread or parachain), auction information
+		/// and the auction deposit are switched.
+		fn swap(origin, #[compact] other: ParaId) {
+			let id = parachains::ensure_parachain(<T as Trait>::Origin::from(origin))?;
+
+			if PendingSwap::get(other) == Some(id) {
+				// actually do the swap.
+				T::OnSwap::ensure_can_swap(id, other)?;
+
```

Comment:
```suggestion

// Remove intention to swap.
PendingSwap::remove(other);
```

This cleans up the "intention to swap" of a para which has successfully done a swap.

I updated a test to check for this, and this will make that test pass.

Reply:
right - without this there is a storage leak